src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -758,10 +758,26 @@ void executeSuppressionActions(IEnumerable<Diagnostic> reportedDiagnostics, bool
                 var suppressors = this.Analyzers.OfType<DiagnosticSuppressor>();
                 if (concurrent)
                 {
-                    Parallel.ForEach(suppressors, suppressor =>
+                    // Kick off tasks to concurrently execute suppressors.
+                    // Note that we avoid using Parallel.ForEach here to avoid wrapped exceptions.
+                    // See https://github.com/dotnet/roslyn/issues/41713 for details.
+                    var tasks = ArrayBuilder<Task>.GetInstance();
+                    try
                     {
-                        AnalyzerExecutor.ExecuteSuppressionAction(suppressor, getSuppressableDiagnostics(suppressor));
-                    });
+                        foreach (var suppressor in suppressors)
+                        {
+                            var task = Task.Run(
+                                () => AnalyzerExecutor.ExecuteSuppressionAction(suppressor, getSuppressableDiagnostics(suppressor)),
+                                AnalyzerExecutor.CancellationToken);
+                            tasks.Add(task);
+                        }
+
+                        Task.WaitAll(tasks.ToArray(), AnalyzerExecutor.CancellationToken);
```

Comment:
📝 I believe `Task.WaitAll` has exactly the same exception behavior as `Parallel.ForEach`.

Reply:
I verified the added test fails prior to this change - the invoked code was throwing an `AggregateException`, but now throws `OperationCanceledException`.

Reply:
Tagging @stephentoub to confirm the semantics of `Task.WaitAll`. I see under the debugger that `OperationCanceledException` thrown from any of tasks is propagated as `OperationCanceledException` itself, not wrapped within an `AggregateException`. We'd still like to know if this is indeed a guarantee from the API, not just a specific instance where this behavior is seen on cancellation.

Reply:
Yes, assuming the wait itself was canceled and the only tasks that didn't complete successfully were canceled:
https://github.com/dotnet/runtime/blob/4807684b13d473c19121fbe757296e7607f3cabf/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L4727-L4735

Reply:
Thanks @stephentoub.

@sharwell Are you fine to sign off on the changes?